sql/partitioning/partition_base.cc

```diff
@@ -0,0 +1,5823 @@
+  m_is_sub_partitioned= m_part_info->is_sub_partitioned();
+  m_is_clone_of= clone_arg;
+  m_clone_mem_root= clone_mem_root_arg;
+  part_share= clone_arg->part_share;
+  m_tot_parts= clone_arg->m_tot_parts;
+  m_pkey_is_clustered= clone_arg->primary_key_is_clustered();
+  DBUG_VOID_RETURN;
+}
+
+/*
+  Initialize handler object
+
+  SYNOPSIS
+    init_handler_variables()
+
+  RETURN VALUE
+    NONE
+*/
+
+void Partition_base::init_handler_variables()
+{
+  active_index= MAX_KEY;
+  m_mode= 0;
+  m_open_test_lock= 0;
+  m_file= NULL;
+  m_file_tot_parts= 0;
+  m_tot_parts= 0;
+  m_pkey_is_clustered= 0;
+  m_myisam= FALSE;
+  m_innodb= FALSE;
+  m_extra_cache= FALSE;
+  m_extra_cache_size= 0;
+  m_extra_prepare_for_update= FALSE;
+  m_extra_cache_part_id= NO_CURRENT_PART_ID;
+  m_handler_status= handler_not_initialized;
+  m_low_byte_first= 1;
+  m_part_func_monotonicity_info= NON_MONOTONIC;
+  /*
+    this allows blackhole to work properly
+  */
+  m_num_locks= 0;
+  m_is_clone_of= NULL;
+  m_clone_mem_root= NULL;
+  part_share= NULL;
+  m_new_partitions_share_refs.empty();
+  m_part_ids_sorted_by_num_of_records= NULL;
+  m_new_file= NULL;
+  m_num_new_partitions= 0;
+  m_indexes_are_disabled= false;
+}
+
+
+const char *Partition_base::table_type() const
+{
+  // we can do this since we only support a single engine type
+  return m_file[0]->table_type();
+}
+
+
+/*
+  Destructor method
+
+  SYNOPSIS
+    ~Partition_base()
+
+  RETURN VALUE
+    NONE
+*/
+
+Partition_base::~Partition_base()
+{
+  DBUG_ENTER("Partition_base::~Partition_base()");
+  if (m_new_partitions_share_refs.elements)
+    m_new_partitions_share_refs.delete_elements();
+  if (m_file != NULL)
+  {
+    uint i;
+    for (i= 0; i < m_tot_parts; i++)
+      delete m_file[i];
+  }
+  my_free(m_part_ids_sorted_by_num_of_records);
+
+  DBUG_VOID_RETURN;
+}
+
+bool Partition_base::init_with_fields()
+{
+  /* Partition info has not yet been initialized, just return true */
+  if (m_handler_status != handler_initialized)
+    return false;
+  /* Pass the call to each partition */
+  for (uint i= 0; i < m_tot_parts; i++)
+  {
+    if (m_file[i]->init_with_fields())
+      return true;
+  }
+  /* Re-read table flags in case init_with_fields caused it to change */
+  cached_table_flags=
+      (m_file[0]->ha_table_flags() & ~(PARTITION_DISABLED_TABLE_FLAGS)) |
+      PARTITION_ENABLED_TABLE_FLAGS;
+  return false;
+}
+
+template <typename Fn>
+bool Partition_base::foreach_partition(const Fn &fn)
+{
+  DBUG_ASSERT(m_part_info);
+  List_iterator_fast<partition_element> part_it(m_part_info->partitions);
+  for (uint i= 0; i < m_part_info->num_parts; ++i)
+  {
+    partition_element *part_elem= part_it++;
+    if (m_is_sub_partitioned)
+    {
+      List_iterator_fast<partition_element> sub_it(part_elem->subpartitions);
+      for (uint j= 0; j < m_part_info->num_subparts; ++j)
+      {
+        partition_element *sub_part_elem= sub_it++;
+        if (!fn(part_elem, sub_part_elem))
```

Comment:
Pass const references, not const pointers, unless they can be legally NULL

Reply:
I agree, but the first parameter can be NULL, the second one can't be NULL. It is not so essentially for me, I can fix if you insist. But I followed the principle of uniformity, it would look strange if two parameters of the same type and meaning were declared differently. What do you think?

Reply:
I agree, OK to leave both as is